source/extensions/filters/http/geoip/geoip_filter.cc

```diff
@@ -0,0 +1,157 @@
+  lookupGeolocationAnonData(config_->geoAnonVpnHeader(), remote_address,
+                            &Driver::getIsAnonymousVpn);
+  lookupGeolocationAnonData(config_->geoAnonHostingHeader(), remote_address,
+                            &Driver::getIsAnonymousHostingProvider);
+  lookupGeolocationAnonData(config_->geoAnonTorHeader(), remote_address,
+                            &Driver::getIsAnonymousTorExitNode);
+  lookupGeolocationAnonData(config_->geoAnonProxyHeader(), remote_address,
+                            &Driver::getIsAnonymousPublicProxy);
+  all_lookups_invoked_ = true;
+
+  // Stop the iteration for headers for the current filter and the filters following.
+  return allLookupsComplete() ? Http::FilterHeadersStatus::Continue
+                              : Http::FilterHeadersStatus::StopIteration;
+}
+
+void GeoipFilter::lookupGeolocationData(
+    const absl::optional<std::string>& geo_header,
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                         const LookupCallbacks& callbacks,
+                         const absl::optional<std::string>& geo_header) const) {
+  if (geo_header) {
+    ++num_lookups_to_perform_;
+    ((*driver_).*func)(remote_address, *this, geo_header);
+  }
+}
+
+void GeoipFilter::lookupGeolocationAnonData(
+    const absl::optional<std::string>& geo_header,
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                         const LookupCallbacks& callbacks,
+                         const absl::optional<std::string>& geo_header) const) {
+  if (geo_header) {
+    ++num_lookups_to_perform_;
+    ((*driver_).*func)(remote_address, *this, geo_header);
+  }
+}
+
+Http::FilterDataStatus GeoipFilter::decodeData(Buffer::Instance&, bool) {
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus GeoipFilter::decodeTrailers(Http::RequestTrailerMap&) {
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void GeoipFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+void GeoipFilter::onLookupComplete(const absl::optional<std::string>& lookup_result,
+                                   const absl::optional<std::string>& geo_header) {
+  if (!request_headers_) {
+    ENVOY_LOG(debug, "Geoip filter: no header to decorate in reguest");
+    return;
+  }
+  ASSERT(num_lookups_to_perform_ > 0);
+  --num_lookups_to_perform_;
+  if (lookup_result) {
+    request_headers_->setCopy(Http::LowerCaseString(geo_header.value()), lookup_result.value());
+    config_->incHit(geo_header.value());
+  }
+  config_->incTotal(geo_header.value());
+  if (allLookupsComplete()) {
+    ENVOY_LOG(debug, "Geoip filter: finished decoding geolocation headers");
+    decoder_callbacks_->continueDecoding();
```

Comment:
I think this doesn't work great with asynchronous callbacks - num_lookups_to_perform_ being updated by the callback directly means it would need to be mutex-guarded or atomic to prevent overlapping operations occurring on different threads. There's also an issue that the filter 'this' can potentially be destroyed while an operation is in flight.

You could resolve this (and other raciness issues so you'd no longer need `all_lookups_invoked_`) by having the callback go through the dispatcher on decoder_callbacks_, and capture a weak_ptr to something owned by the filter. Using the dispatcher effectively makes the filter single-threaded again (i.e. the filter part of the callback can't be called while decodeHeaders is still processing), and capturing a weak_ptr means you can abandon the rest of the callback if the filter has been destroyed before the callback is called.

I don't think there's a way to do this safely where the driver takes only a reference to the callback object - it must take a shared_ptr or a std::function so that it's safe for the filter to be destroyed without causing the driver's async callback to lose its necessary context.

My suggestion would be to pass the driver a unique_ptr to an additional LookupCallbacks implementation that contains a reference to the dispatcher and a weak_ptr that leads back to the filter. When called, that object could then `dispatcher_->dispatch([the_weak_ptr, callback_params]() {if (auto p = the_weak_ptr.lock()) p->onLookupComplete(callback_params);})`

Reply:
I may have misinterpreted the docs, was thinking that http filter instance is created per stream/request and that no guarding is needed.

Reply:
No, you have that part right. The issue is that the driver isn't part of the filter, and (being asynchronous) may have its own thread. So now there's the filter's thread and the driver's thread (or threads! There's multiple calls to the driver in flight at the same time), both/all touching about the same variables. If the callbacks are applied through the dispatcher then they end up back on the filter's thread which therefore would eliminate any need to guard the variables. (But does still need to be careful that the filter instance hasn't been destroyed by the time the dispatcher gets around to dispatching.)

Reply:
Yeap, fair. Thanks @ravenblackx and apologies for delay, I have been OOO entire February. Am now catching up and continuing this work.